overlord/snapstate/handlers.go

```diff
@@ -898,38 +898,44 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 // asyncRefreshOnSnapClose asynchronously waits for the snap the close, notifies
 // the user and then triggers an auto-refresh.
 func asyncRefreshOnSnapClose(st *state.State, refreshInfo *userclient.PendingSnapRefreshInfo) error {
-	var monitoredSnaps map[string]bool
+	var monitoredSnaps map[string]chan<- bool
 	if cachedMonitored := st.Cached("monitored-snaps"); cachedMonitored != nil {
-		monitoredSnaps, _ = cachedMonitored.(map[string]bool)
+		monitoredSnaps, _ = cachedMonitored.(map[string]chan<- bool)
 	} else {
-		monitoredSnaps = make(map[string]bool)
+		monitoredSnaps = make(map[string]chan<- bool)
 	}
 
 	// there's already a goroutine waiting for this snap to close so just notify
-	if monitoredSnaps[refreshInfo.InstanceName] {
+	if monitoredSnaps[refreshInfo.InstanceName] != nil {
 		asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
 		return nil
 	}
 
 	// monitor the snap until it closes
-	done := make(chan string)
+	done := make(chan string, 1)
 	if err := cgroupMonitorSnapEnded(refreshInfo.InstanceName, done); err != nil {
 		return fmt.Errorf("cannot monitor for snap closure: %w", err)
 	}
 
 	// notify the user about the blocked refresh
 	asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
 
-	monitoredSnaps[refreshInfo.InstanceName] = true
+	abort := make(chan bool, 1)
+	monitoredSnaps[refreshInfo.InstanceName] = abort
 	st.Cache("monitored-snaps", monitoredSnaps)
 
-	// TODO: clear the monitoring related state and goroutines when an auto-refresh happens
 	go func() {
-		<-done
+		var continueAutoRefresh bool
```

Comment:
nitpick: in this case this might be a bit clearer:

continueAutoRefresh := false